File: evidence/ores-cli-pr63/go/main.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"go/ast"
	"go/parser"
	"go/token"
	"os"
	"path"
	"path/filepath"
	"regexp"
	"runtime"
	"sort"
	"strconv"
	"strings"
)

const schemaVersion = "ores.code-config-audit.v1"
const maxFiles = 4096
const maxBytes int64 = 64 * 1024 * 1024
const maxFileBytes int64 = 2 * 1024 * 1024

var skipDirs = map[string]bool{
	".git": true, ".dart_tool": true, ".idea": true, ".next": true,
	".turbo": true, ".zed": true, "build": true, "coverage": true,
	"dist": true, "generated": true, "node_modules": true, "target": true,
	"vendor": true, "zed_modules": true,
}

var regexPatterns = []struct {
	kind string
	re   *regexp.Regexp
}{
	{"config-literal", regexp.MustCompile(`["` + "`" + `]([^"` + "`" + `\n]*\.toml)["` + "`" + `]`)},
	{"env-api", regexp.MustCompile(`\b(os\.(?:Getenv|LookupEnv))\b`)},
	{"toml-api", regexp.MustCompile(`\b(toml\.(?:Decode|Unmarshal|NewDecoder))\b`)},
}

type Event struct {
	Kind   string `json:"kind"`
	Value  string `json:"value,omitempty"`
	Line   uint32 `json:"line"`
	Column uint32 `json:"column"`
}

type FileReceipt struct {
	Path        string  `json:"path"`
	SyntaxValid bool    `json:"syntaxValid"`
	RegexHits   []Event `json:"regexHits"`
	ASTEvents   []Event `json:"astEvents"`
}

type Receipt struct {
	SchemaVersion string        `json:"schemaVersion"`
	Language      string        `json:"language"`
	Parser        string        `json:"parser"`
	ParserVersion string        `json:"parserVersion"`
	Files         []FileReceipt `json:"files"`
}

func main() {
	rootFlag := flag.String("root", "", "repository root")
	formatFlag := flag.String("format", "", "output format")
	flag.Parse()
	if *rootFlag == "" || *formatFlag != "json" || flag.NArg() != 0 {
		fatal("usage: ores-ast-go --root <path> --format json")
	}
	root, err := filepath.EvalSymlinks(*rootFlag)
	if err != nil {
		fatal("root could not be resolved")
	}
	root, err = filepath.Abs(root)
	if err != nil {
		fatal("root could not be made absolute")
	}

	files, err := inventory(root)
	if err != nil {
		fatal(err.Error())
	}
	fset := token.NewFileSet()
	receipts := make([]FileReceipt, 0, len(files))
	for _, file := range files {
		receipts = append(receipts, inspectFile(root, file, fset))
	}
	sort.Slice(receipts, func(i, j int) bool { return receipts[i].Path < receipts[j].Path })

	encoded, err := json.Marshal(Receipt{
		SchemaVersion: schemaVersion,
		Language:      "go",
		Parser:        "go/parser+go/ast",
		ParserVersion: runtime.Version(),
		Files:         receipts,
	})
	if err != nil {
		fatal("receipt serialization failed")
	}
	fmt.Println(string(encoded))
}

func fatal(message string) {
	fmt.Fprintln(os.Stderr, message)
	os.Exit(2)
}

func inventory(root string) ([]string, error) {
	files := make([]string, 0)
	var total int64
	err := filepath.WalkDir(root, func(filePath string, entry os.DirEntry, walkErr error) error {
		if walkErr != nil {
			return walkErr
		}
		if entry.Type()&os.ModeSymlink != 0 {
			if entry.IsDir() {
				return filepath.SkipDir
			}
			return nil
		}
		if entry.IsDir() {
			if filePath != root && skipDirs[entry.Name()] {
				return filepath.SkipDir
			}
			return nil
		}
		if filepath.Ext(entry.Name()) != ".go" {
			return nil
		}
		info, err := entry.Info()
		if err != nil {
			return err
		}
		if info.Size() > maxFileBytes {
			return fmt.Errorf("source file exceeds bound")
		}
		total += info.Size()
		if total > maxBytes || len(files) >= maxFiles {
			return fmt.Errorf("source inventory exceeds bound")
		}
		files = append(files, filePath)
		return nil
	})
	if err != nil {
		return nil, err
	}
	sort.Strings(files)
	return files, nil
}

func inspectFile(root, file string, fset *token.FileSet) FileReceipt {
	textBytes, err := os.ReadFile(file)
	if err != nil {
		fatal("source file could not be read")
	}
	text := string(textBytes)
	parsed, parseErr := parser.ParseFile(fset, file, text, parser.AllErrors)
	result := FileReceipt{
		Path:        relative(root, file),
		SyntaxValid: parseErr == nil,
		RegexHits:   regexLane(text),
		ASTEvents:   []Event{},
	}
	if parsed != nil {
		result.ASTEvents = astLane(parsed, fset)
	}
	return result
}

func regexLane(text string) []Event {
	events := make([]Event, 0)
	lineStarts := []int{0}
	for i, r := range text {
		if r == '\n' {
			lineStarts = append(lineStarts, i+1)
		}
	}
	for _, pattern := range regexPatterns {
		for _, match := range pattern.re.FindAllStringSubmatchIndex(text, -1) {
			start := match[0]
			value := text[match[2]:match[3]]
			line, column := positionFromOffset(lineStarts, start)
			events = append(events, Event{Kind: pattern.kind, Value: value, Line: line, Column: column})
		}
	}
	return dedupe(events)
}

func importAliases(file *ast.File) map[string]string {
	aliases := make(map[string]string)
	for _, spec := range file.Imports {
		importPath, err := strconv.Unquote(spec.Path.Value)
		if err != nil || importPath == "" {
			continue
		}
		name := path.Base(importPath)
		if spec.Name != nil {
			name = spec.Name.Name
		}
		if name == "_" || name == "." || name == "" {
			continue
		}
		aliases[name] = importPath
	}
	return aliases
}

func isTomlImport(importPath string) bool {
	return importPath == "github.com/BurntSushi/toml" ||
		strings.Contains(importPath, "pelletier/go-toml") ||
		strings.HasSuffix(importPath, "/toml")
}

func astLane(file *ast.File, fset *token.FileSet) []Event {
	events := make([]Event, 0)
	aliases := importAliases(file)
	ast.Inspect(file, func(node ast.Node) bool {
		call, ok := node.(*ast.CallExpr)
		if !ok {
			return true
		}
		sel, ok := call.Fun.(*ast.SelectorExpr)
		if !ok {
			return true
		}
		pkg, ok := sel.X.(*ast.Ident)
		if !ok {
			return true
		}
		pos := fset.Position(call.Pos())
		pkgPath := aliases[pkg.Name]
		isOS := pkg.Name == "os" || pkgPath == "os"
		isToml := pkg.Name == "toml" || isTomlImport(pkgPath)
		if isOS && (sel.Sel.Name == "Getenv" || sel.Sel.Name == "LookupEnv") {
			events = append(events, Event{
				Kind: "env-read", Value: firstString(call.Args),
				Line: uint32(pos.Line), Column: uint32(pos.Column),
			})
		}
		if isOS && (sel.Sel.Name == "ReadFile" || sel.Sel.Name == "Open") {
			if value := firstString(call.Args); strings.HasSuffix(value, ".toml") {
				events = append(events, Event{
					Kind: "config-read", Value: value,
					Line: uint32(pos.Line), Column: uint32(pos.Column),
				})
			}
		}
		if isToml && (sel.Sel.Name == "Decode" || sel.Sel.Name == "Unmarshal" || sel.Sel.Name == "NewDecoder") {
			events = append(events, Event{
				Kind: "toml-parse", Line: uint32(pos.Line), Column: uint32(pos.Column),
			})
		}
		return true
	})
	return dedupe(events)
}

func firstString(args []ast.Expr) string {
	if len(args) == 0 {
		return ""
	}
	literal, ok := args[0].(*ast.BasicLit)
	if !ok || literal.Kind != token.STRING {
		return ""
	}
	value, err := strconv.Unquote(literal.Value)
	if err != nil {
		return ""
	}
	return value
}

func relative(root, file string) string {
	rel, err := filepath.Rel(root, file)
	if err != nil || rel == "." || strings.HasPrefix(rel, ".."+string(filepath.Separator)) || filepath.IsAbs(rel) {
		fatal("source escaped root")
	}
	return filepath.ToSlash(rel)
}

func positionFromOffset(lineStarts []int, offset int) (uint32, uint32) {
	lineIndex := sort.Search(len(lineStarts), func(i int) bool { return lineStarts[i] > offset }) - 1
	if lineIndex < 0 {
		lineIndex = 0
	}
	return uint32(lineIndex + 1), uint32(offset - lineStarts[lineIndex] + 1)
}

func dedupe(events []Event) []Event {
	seen := make(map[string]bool)
	result := make([]Event, 0, len(events))
	for _, event := range events {
		key := fmt.Sprintf("%s\x00%s\x00%d\x00%d", event.Kind, event.Value, event.Line, event.Column)
		if seen[key] {
			continue
		}
		seen[key] = true
		result = append(result, event)
	}
	sort.Slice(result, func(i, j int) bool {
		if result[i].Line != result[j].Line {
			return result[i].Line < result[j].Line
		}
		if result[i].Column != result[j].Column {
			return result[i].Column < result[j].Column
		}
		return result[i].Kind < result[j].Kind
	})
	return result
}

File: evidence/ores-cli-pr63/rust/src/main.rs
use std::collections::BTreeSet;
use std::fs;
use std::path::{Path, PathBuf};

use proc_macro2::Span;
use regex::Regex;
use serde::Serialize;
use syn::spanned::Spanned;
use syn::visit::Visit;
use syn::{Expr, ExprCall, Lit, Macro, Path as SynPath};

const SCHEMA: &str = "ores.code-config-audit.v1";
const MAX_FILES: usize = 4_096;
const MAX_BYTES: u64 = 64 * 1024 * 1024;
const MAX_FILE_BYTES: u64 = 2 * 1024 * 1024;

const SKIP_DIRS: &[&str] = &[
    ".git",
    ".dart_tool",
    ".idea",
    ".next",
    ".turbo",
    ".zed",
    "build",
    "coverage",
    "dist",
    "generated",
    "node_modules",
    "target",
    "vendor",
    "zed_modules",
];

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize)]
#[serde(rename_all = "camelCase")]
struct Event {
    kind: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    value: Option<String>,
    line: u32,
    column: u32,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
struct FileReceipt {
    path: String,
    syntax_valid: bool,
    regex_hits: Vec<Event>,
    ast_events: Vec<Event>,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
struct Receipt {
    schema_version: &'static str,
    language: &'static str,
    parser: &'static str,
    parser_version: &'static str,
    files: Vec<FileReceipt>,
}

fn main() {
    match run() {
        Ok(receipt) => match serde_json::to_string(&receipt) {
            Ok(json) => println!("{json}"),
            Err(_) => fail("receipt serialization failed"),
        },
        Err(message) => fail(&message),
    }
}

fn run() -> Result<Receipt, String> {
    let root = parse_args()?;
    let root = fs::canonicalize(root).map_err(|_| "root could not be resolved".to_owned())?;
    if !root.is_dir() {
        return Err("root must be a directory".to_owned());
    }

    let files = inventory(&root)?;
    let mut receipts = Vec::with_capacity(files.len());
    for file in files {
        receipts.push(inspect_file(&root, &file)?);
    }
    receipts.sort_by(|left, right| left.path.cmp(&right.path));

    Ok(Receipt {
        schema_version: SCHEMA,
        language: "rust",
        parser: "syn",
        parser_version: "3.0.5",
        files: receipts,
    })
}

fn parse_args() -> Result<PathBuf, String> {
    let mut args = std::env::args().skip(1);
    let mut root = None;
    let mut format = None;
    while let Some(arg) = args.next() {
        let value = args
            .next()
            .ok_or_else(|| "every option requires one value".to_owned())?;
        match arg.as_str() {
            "--root" => root = Some(PathBuf::from(value)),
            "--format" => format = Some(value),
            _ => return Err("unknown adapter option".to_owned()),
        }
    }
    if format.as_deref() != Some("json") {
        return Err("only --format json is supported".to_owned());
    }
    root.ok_or_else(|| "--root is required".to_owned())
}

fn inventory(root: &Path) -> Result<Vec<PathBuf>, String> {
    fn walk(
        root: &Path,
        dir: &Path,
        files: &mut Vec<PathBuf>,
        bytes: &mut u64,
    ) -> Result<(), String> {
        let mut entries = fs::read_dir(dir)
            .map_err(|_| "source directory could not be read".to_owned())?
            .collect::<Result<Vec<_>, _>>()
            .map_err(|_| "source directory entry could not be read".to_owned())?;
        entries.sort_by_key(|entry| entry.file_name());

        for entry in entries {
            let path = entry.path();
            let metadata = fs::symlink_metadata(&path)
                .map_err(|_| "source metadata could not be read".to_owned())?;
            if metadata.file_type().is_symlink() {
                continue;
            }
            if metadata.is_dir() {
                let name = entry.file_name();
                let name = name.to_string_lossy();
                if path != root && SKIP_DIRS.contains(&name.as_ref()) {
                    continue;
                }
                walk(root, &path, files, bytes)?;
                continue;
            }
            if !metadata.is_file()
                || path.extension().and_then(|value| value.to_str()) != Some("rs")
            {
                continue;
            }
            if metadata.len() > MAX_FILE_BYTES {
                return Err("source file exceeds bound".to_owned());
            }
            *bytes = bytes.saturating_add(metadata.len());
            if *bytes > MAX_BYTES || files.len() >= MAX_FILES {
                return Err("source inventory exceeds bound".to_owned());
            }
            files.push(path);
        }
        Ok(())
    }

    let mut files = Vec::new();
    let mut bytes = 0;
    walk(root, root, &mut files, &mut bytes)?;
    files.sort();
    Ok(files)
}

fn inspect_file(root: &Path, file: &Path) -> Result<FileReceipt, String> {
    let source = fs::read_to_string(file).map_err(|_| "source file is not UTF-8".to_owned())?;
    let relative = relative(root, file)?;
    let regex_hits = regex_lane(&source)?;
    match syn::parse_file(&source) {
        Ok(parsed) => {
            let mut visitor = ConfigVisitor::new();
            visitor.visit_file(&parsed);
            Ok(FileReceipt {
                path: relative,
                syntax_valid: true,
                regex_hits,
                ast_events: visitor.finish(),
            })
        }
        Err(_) => Ok(FileReceipt {
            path: relative,
            syntax_valid: false,
            regex_hits,
            ast_events: Vec::new(),
        }),
    }
}

fn regex_lane(source: &str) -> Result<Vec<Event>, String> {
    let config = Regex::new(r#"[\"']([^\"'\n]*\.toml)[\"']"#)
        .map_err(|_| "config regex did not compile".to_owned())?;
    let env = Regex::new(
        r"\b(?:std::env::(?:var|var_os)|env::(?:var|var_os)|env!|option_env!)\b",
    )
    .map_err(|_| "env regex did not compile".to_owned())?;
    let toml = Regex::new(r"\btoml::(?:from_str|from_slice|de::from_str)\b")
        .map_err(|_| "TOML regex did not compile".to_owned())?;

    let mut events = BTreeSet::new();
    for captures in config.captures_iter(source) {
        let Some(value) = captures.get(1) else {
            continue;
        };
        let (line, column) = line_column(source, value.start());
        events.insert(Event {
            kind: "config-literal".to_owned(),
            value: Some(value.as_str().to_owned()),
            line,
            column,
        });
    }
    for found in env.find_iter(source) {
        let (line, column) = line_column(source, found.start());
        events.insert(Event {
            kind: "env-api".to_owned(),
            value: Some(found.as_str().to_owned()),
            line,
            column,
        });
    }
    for found in toml.find_iter(source) {
        let (line, column) = line_column(source, found.start());
        events.insert(Event {
            kind: "toml-api".to_owned(),
            value: Some(found.as_str().to_owned()),
            line,
            column,
        });
    }
    Ok(events.into_iter().collect())
}

struct ConfigVisitor {
    events: BTreeSet<Event>,
}

impl ConfigVisitor {
    fn new() -> Self {
        Self {
            events: BTreeSet::new(),
        }
    }

    fn finish(self) -> Vec<Event> {
        self.events.into_iter().collect()
    }

    fn push(&mut self, kind: &str, value: Option<String>, span: Span) {
        let start = span.start();
        self.events.insert(Event {
            kind: kind.to_owned(),
            value,
            line: u32::try_from(start.line).unwrap_or(u32::MAX),
            column: u32::try_from(start.column.saturating_add(1)).unwrap_or(u32::MAX),
        });
    }
}

impl<'ast> Visit<'ast> for ConfigVisitor {
    fn visit_expr_call(&mut self, node: &'ast ExprCall) {
        if let Expr::Path(function) = node.func.as_ref() {
            let function_name = path_name(&function.path);
            let first = first_string_argument(node);
            if matches!(
                function_name.as_str(),
                "std::env::var" | "std::env::var_os" | "env::var" | "env::var_os"
            ) {
                self.push("env-read", first, node.span());
            }
            if matches!(
                function_name.as_str(),
                "std::fs::read_to_string"
                    | "std::fs::read"
                    | "fs::read_to_string"
                    | "fs::read"
            ) && first.as_deref().is_some_and(|value| value.ends_with(".toml"))
            {
                self.push("config-read", first, node.span());
            }
            if matches!(
                function_name.as_str(),
                "toml::from_str" | "toml::from_slice" | "toml::de::from_str"
            ) {
                self.push("toml-parse", None, node.span());
            }
        }
        syn::visit::visit_expr_call(self, node);
    }

    fn visit_macro(&mut self, node: &'ast Macro) {
        let name = path_name(&node.path);
        if matches!(name.as_str(), "include_str" | "include_bytes") {
            if let Ok(value) = syn::parse2::<syn::LitStr>(node.tokens.clone()) {
                let value = value.value();
                if value.ends_with(".toml") {
                    self.push("config-read", Some(value), node.span());
                }
            }
        }
        if matches!(name.as_str(), "env" | "option_env") {
            if let Ok(value) = syn::parse2::<syn::LitStr>(node.tokens.clone()) {
                self.push("env-read", Some(value.value()), node.span());
            }
        }
        syn::visit::visit_macro(self, node);
    }
}

fn first_string_argument(node: &ExprCall) -> Option<String> {
    match node.args.first()? {
        Expr::Lit(literal) => match &literal.lit {
            Lit::Str(value) => Some(value.value()),
            _ => None,
        },
        _ => None,
    }
}

fn path_name(path: &SynPath) -> String {
    path.segments
        .iter()
        .map(|segment| segment.ident.to_string())
        .collect::<Vec<_>>()
        .join("::")
}

fn line_column(source: &str, offset: usize) -> (u32, u32) {
    let prefix = &source[..offset.min(source.len())];
    let line = prefix.bytes().filter(|byte| *byte == b'\n').count() + 1;
    let column = prefix
        .rsplit_once('\n')
        .map_or(prefix.len() + 1, |(_, tail)| tail.len() + 1);
    (
        u32::try_from(line).unwrap_or(u32::MAX),
        u32::try_from(column).unwrap_or(u32::MAX),
    )
}

fn relative(root: &Path, file: &Path) -> Result<String, String> {
    let relative = file
        .strip_prefix(root)
        .map_err(|_| "source escaped root".to_owned())?;
    let value = relative.to_string_lossy().replace('\\', "/");
    if value.is_empty() || value.starts_with("../") || value.starts_with('/') {
        return Err("source escaped root".to_owned());
    }
    Ok(value)
}

fn fail(message: &str) -> ! {
    eprintln!("{message}");
    std::process::exit(2)
}

#[cfg(test)]
mod tests {
    use syn::visit::Visit;

    use super::ConfigVisitor;

    #[test]
    fn repeated_identical_env_reads_keep_distinct_ast_positions() {
        let source = r#"
fn load() {
    let _ = std::env::var("REDIS_URL");
    let _ = std::env::var("REDIS_URL");
}
"#;
        let parsed = syn::parse_file(source).expect("valid Rust fixture");
        let mut visitor = ConfigVisitor::new();
        visitor.visit_file(&parsed);
        let events = visitor
            .finish()
            .into_iter()
            .filter(|event| event.kind == "env-read" && event.value.as_deref() == Some("REDIS_URL"))
            .collect::<Vec<_>>();

        assert_eq!(events.len(), 2);
        assert_eq!(events[0].line, 3);
        assert_eq!(events[1].line, 4);
        assert_ne!((events[0].line, events[0].column), (events[1].line, events[1].column));
    }
}
